### PR TITLE
Add cmock and unity support

### LINKED ISSUE #30391: Unit Testing in Zephyr
## Introduction

While testing the modules written for Zephyr, I noticed that there is no environment to create Unit Tests. Currently, it is not possible to test a module isolated from others. This problem affects both Zephyr developers and those who develop applications based on it.

### Problem description

The "z_test" does not allow to create of module tests in isolation from included modules.
The "z_test" does not provide any mock functionality.

### Proposed change

Integrate Unity and Cmock with Zephyr. 
In the next step maybe we could add unit tests to CI? Unity allows for easy generation of test results into XML format.

## Detailed RFC

Proposal:
https://github.com/jakub-uC/zephyr/pull/1

### Dependencies

In my opinion, it is only added value and will not impact existing projects.

### Concerns and Unresolved Questions

Maintaining z_test and unit test framework at the same time

## Alternatives

Keep using z_test only. Some other vendors like Nordic will keep Unity and Cmock on their Zephyr fork.

### PR DESCRIPTION
**Pull request by mkschreder**
## What was done

This MR adds cmock and unity support including a custom python script that uses clang cindex to parse the headers that will be mocked and extract necessary data for mocking the functions within.

## Why is this important

This adds ability to do full 100% branch and line testing for ANY C file. Zephyr unit testing currently is very limited to what can be tested. But using this approach 100% line and branch coverage can be easily achieved. It is not strictly unit testing because we still need to link it with zephyr platform but it is much much better than any other approach I have used so far.

Using this approach we can test every static private method in any part of the system. 

## How good is the support

- kernel.h can be mocked in full
- gpio driver interface can be mocked in full
- spi interface can be mocked in full
- ... probably many more, but these are the ones I have tried so far.

You can do the following:

- Mock any function for which you can provide a header file
- Specify constraints for order of calls being made (if you enable strict order checking in cmock config)
- Specify expected arguments and return value for sequences of calls
- Specify custom callbacks for mocked functions where you can implement device specific logic (example: https://gitlab.com/swedishembedded/platform/sdk/-/blob/a42214fd59ecb07f3863248dd15d6b01b4e52b7e/tests/drivers/gpio/mcp23s17/src/unit.c)

See cmock documentation for more details: https://github.com/ThrowTheSwitch/CMock/blob/master/docs/CMock_Summary.md

Other kernel headers will have to be handled individually and support *may* need to be added to the cleanup script so that cleaned up headers can be provided to cmock.

## Example

Unit test and integration test are combined into one. This may seem clunky at first, but it avoids a lot of duplication. Integration test is run by default. You can configure two targets in your sample.yaml file that set CMOCK to y for one and drop it for the other. 

Build and run unit test:

`west build -p -b native_posix tests/cmock/example -- -DCONFIG_CMOCK=y`

Build and run integration test:

`west build -p -b native_posix tests/cmock/example
`
## Questions

- **Does unity work on all platforms?** Yes there are no limitations for unity. 
- **Does CMock work on all platforms?** Yes, but you need a lot of RAM because it uses large RAM buffers for setting up expectations. Thus it works best on native_posix (where you will be running majority of your tests anyway). 

## Known issues

Cmock in its current state crashes with latest zephyr repo. This needs to be addressed in the main cmock repository. This is why this MR is targeted at Zephyr 3.0.0 for now. 

Fixes #30391

Signed-off-by: Martin Schröder <mkschreder.uk@gmail.com>

### CODE REVIEW

**Comment by carlescufi** [comment]
I thought that we were settling on FFF instead of Unity/CMock. CC @yperess

**Comment by nashif** [comment]
> I thought that we were settling on FFF instead of Unity/CMock.

for in tree tests, yes, but we should allow support for additional frameworks for someone to write out of tree tests for their applications and use cases, as long as it does not interfere...

**Comment by stephanosio** [comment]
Converting to draft. This will not work until it is updated such that it can be applied onto the `main` branch.

Also please review the [Contribution Guidelines](https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-guidelines) and update the commit message; also, this PR should be split into more manageable commits.

For the external modules, please review the following document:
https://docs.zephyrproject.org/latest/contribute/external.html#submission-and-review-process

**Comment by cfriedt** [comment]
I wouldn't mind seeing this cleaned up and ready to merge into main as well. It's quite useful and it can be a healthy alternative

**Comment by mkschreder** [comment]
> I wouldn't mind seeing this cleaned up and ready to merge into main as well. It's quite useful and it can be a healthy alternative

It currently works in 3.0.0 but for main there is a crash in cmock that needs to be investigated first. If anyone feels like looking into what changes in zephyr are causing cmock to crash in 3.1 then feel free to look into it.

**Comment by mkschreder** [comment]
> Converting to draft. This will not work until it is updated such that it can be applied onto the `main` branch.
> 
> Also please review the [Contribution Guidelines](https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-guidelines) and update the commit message; also, this PR should be split into more manageable commits.
> 
> For the external modules, please review the following document: https://docs.zephyrproject.org/latest/contribute/external.html#submission-and-review-process

There are a few dependencies like ruby that need to be installed in CI docker image. How do we handle this?

**Comment by mkschreder** [comment]
Update: I have created a standalone project for this here: https://github.com/swedishembedded/testing

Example unit test using this setup is here: https://github.com/swedishembedded/sdk/blob/main/tests/drivers/gpio/mcp23s17/src/unit.c

Cmake file that generates mocks is here: https://github.com/swedishembedded/sdk/blob/main/tests/drivers/gpio/mcp23s17/CMakeLists.txt

All one has to do to start using CMock with zephyr now is to just add it to any zephyr project as a west feed: 

```
remotes:
    - name: swedishembedded
      url-base: https://github.com/swedishembedded
- name: testing
      path: modules/test/testing
      revision: v0.34.0
      remote: swedishembedded
      import: true
```

What we could do is start using it as a west feed for now, stabilize it and then once it is working 100% with all zephyr use cases then it can later be merged into zephyr as well. There could be a benefit to maintain it as a west feed too because I estimate that a lot of zephyr headers will need to have additional "custom cleanup" before they can be mocked easily

Current implementation works rather well. Filter script is here: https://github.com/swedishembedded/testing/blob/main/cmock/gen_headers.py. But it has not been tried on ALL zephyr headers which I think would be a good thing to achieve before an official merge into trunk. It would be nice to also implement a test that tries to generate mock for every single header under include/zephyr..

**Comment by mkschreder** [comment]
ruby is needed on the build system to make this build.

File: subsys/testsuite/cmock/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,128 @@
+# Copyright (c) 2022 Swedish Embedded Consulting Group AB
+# Author: Martin Schröder <martin.schroder@swedishembedded.com>
+# SPDX-License-Identifier: Apache-2.0
+# Consulting: https://swedishembedded.com/go
```
Comment (cfriedt) [inline subsys/testsuite/cmock/CMakeLists.txt:4]: We can probably go without the advertising in Copyright statements.

I thought Zephyr did have a consultant directory at one point that you could sign up to be listed on.

@kestewart - do you have more information on that?

**Review by cfriedt — CHANGES_REQUESTED** [review]

File: subsys/testsuite/cmock/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+get_property(CMOCK_DIR GLOBAL PROPERTY CMOCK_DIR)
+
+# locate cmock config file
+set(CMOCK_CONFIG_FILE
+    ${CMAKE_CURRENT_LIST_DIR}/cmock_cfg.yaml
+    CACHE STRING "")
+
+# check if we have ruby executable
+find_program(RUBY_PATH ruby)
+if(${RUBY_PATH} STREQUAL RUBY_PATH-NOTFOUND)
```
Comment (fkokosinski) [inline subsys/testsuite/cmock/CMakeLists.txt:21]: If we're adding an example (in `tests/cmock/example`) I believe we should have some more or less automated way to get all the required software to run cmock, which this PR is currently lacking.

File: subsys/testsuite/cmock/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+  endforeach()
+  zephyr_link_libraries(${LINKER_OPTS})
+endfunction()
+
+function(cmock_add_linker_wraps HEADER_PATH)
+  set(FUNCTION_LIST_FILE "${HEADER_PATH}.flist")
+
+  execute_process(
+    COMMAND
+      ${PYTHON_EXECUTABLE} ${ZEPHYR_TESTING_MODULE_DIR}/cmock/gen_function_list.py
```
Comment (fkokosinski) [inline subsys/testsuite/cmock/CMakeLists.txt:68]: Why is `ZEPHYR_TESTING_MODULE_DIR` not set by default? The `subsys/testsuite` location seems to be fixed and could therefore be the default value of this variable.

File: subsys/testsuite/cmock/Kconfig [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+# Copyright 2022 Martin Schröder <martin.schroder@swedishembedded.com>
+# Consulting: https://swedishembedded.com/go
+# Training: https://swedishembedded.com/tag/training
+
+config CMOCK
+	bool "Use CMock mocking framework"
+	select TEST
+	depends on UNITY
+	help
+			This option enables CMock testing.
```
Comment (fkokosinski) [inline subsys/testsuite/cmock/Kconfig:11]: Wrong indentation on the help string. Should be 2 spaces.

File: subsys/testsuite/unity/Kconfig [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+# Copyright 2022 Martin Schröder <info@swedishembedded.com>
+# Consulting: https://swedishembedded.com/go
+# Training: https://swedishembedded.com/tag/training
+
+config UNITY
+	bool "Use Unity unit test framework"
+	select TEST
+	depends on !ZTEST
+	help
+			This enables unity test framework which you should use instead of ZTEST.
```
Comment (fkokosinski) [inline subsys/testsuite/unity/Kconfig:11]: Ditto.

**Review by fkokosinski — COMMENTED** [review]
Hi @mkschreder! I tried building the `tests/cmock/example` but it's unclear to me from what you had provided how to do this properly:
```
west build -p -b native_posix tests/cmock/example -- -DCONFIG_CMOCK=y -DZEPHYR_TESTING_MODULE_DIR="$(pwd)/subsys/testsuite"
```
It looks like the Python scripts couldn't find libclang:
```
clang.cindex.LibclangError: libclang-14.so: cannot open shared object file: No such file or directory. To provide a path to libclang use Config.set_library_path() or Config.set_library_file().
```

Note that I'm building this on the Zephyr docker image. Could you describe in more detail how to prepare the environment to be able to build this sample?

File: subsys/testsuite/cmock/CMakeLists.txt (hunk 2 above)
Comment (mkschreder) [reply subsys/testsuite/cmock/CMakeLists.txt:21]: > Note that I'm building this on the Zephyr docker image. Could you describe in more detail how to prepare the environment to be able to build this sample?

You will need these packages: https://github.com/swedishembedded/develop/blob/d6a16063c30d465caf35adc5bc165ad541e80197/Dockerfile.build#L52

And also probably clang pip package: https://github.com/swedishembedded/develop/blob/d6a16063c30d465caf35adc5bc165ad541e80197/Dockerfile.build#L22

Basically the script uses libclang to extract function names from headers.

Comment (mkschreder) [reply subsys/testsuite/cmock/CMakeLists.txt:21]: What would be a standard procedure for making CI pass here? Should I raise a PR at the docker image repo first with the changes or should I simply add missing packages to ci scripts here and then somebody else will add them to the docker image?

**Comment by stephanosio** [comment]
> > I thought that we were settling on FFF instead of Unity/CMock.
> 
> for in tree tests, yes, but we should allow support for additional frameworks for someone to write out of tree tests for their applications and use cases, as long as it does not interfere...

If we have settled on the FFF for in-tree tests, the Unity/CMock support would be better off as a separate independent module instead of an in-tree component -- in fact, the author already has it available in module form: https://github.com/swedishembedded/testing.

Having multiple testing frameworks available in the Zephyr main tree is confusing and can be misleading for developers. Also the fact that CMock is written in Ruby and that it requires a Ruby interpreter does not help either.

File: tests/cmock/example/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,14 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.20.0)
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(test)
+
+if(CONFIG_CMOCK)
+  unity_generate_test_runner(src/unit.c)
+  cmock_generate_mocks(${ZEPHYR_BASE}/include/kernel.h .)
```
Comment (fkokosinski) [inline tests/cmock/example/CMakeLists.txt:9]: This path seems incorrect and CMake can't find this file. The correct path would be `${ZEPHYR_BASE}/include/zephyr/kernel.h`.

After applying this change (and installing stuff you send in the other thread), I get this error message:
```
ValueError: Unknown template argument kind 350
CMake Error at /workdir/zephyr/subsys/testsuite/cmock/CMakeLists.txt:75 (message):
  1
Call Stack (most recent call first):
  /workdir/zephyr/subsys/testsuite/cmock/CMakeLists.txt:122 (cmock_add_linker_wraps)
  CMakeLists.txt:9 (cmock_generate_mocks)


CMake Error at /workdir/zephyr/subsys/testsuite/cmock/CMakeLists.txt:76 (message):
  Error generating function list from /workdir/zephyr/build/mocks/./kernel.h
Call Stack (most recent call first):
  /workdir/zephyr/subsys/testsuite/cmock/CMakeLists.txt:122 (cmock_add_linker_wraps)
  CMakeLists.txt:9 (cmock_generate_mocks)


-- Configuring incomplete, errors occurred!
See also "/workdir/zephyr/build/CMakeFiles/CMakeOutput.log".
```
I can confirm that `zephyr/build/mocks/./kernel.h` exists and is full of kernel symbols and macros, so I guess that the clang part worked correctly.

Comment (mkschreder) [reply tests/cmock/example/CMakeLists.txt:9]: This is because it was originally written for older version of zephyr before the include directory change. The path now is "zephyr/include/zephyr/kernel.h"
